code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm

```diff
@@ -0,0 +1,529 @@
+	return ..()
+
+/obj/machinery/airalarm/examine(mob/user)
+	. = ..()
+	switch(buildstage)
+		if(AIRALARM_BUILD_NO_CIRCUIT)
+			. += span_notice("It is missing air alarm electronics.")
+		if(AIRALARM_BUILD_NO_WIRES)
+			. += span_notice("It is missing wiring.")
+		if(AIRALARM_BUILD_COMPLETE)
+			. += span_notice("Right-click to [locked ? "unlock" : "lock"] the interface.")
+
+/obj/machinery/airalarm/ui_status(mob/user)
+	if(user.has_unlimited_silicon_privilege && aidisabled)
+		to_chat(user, "AI control has been disabled.")
+	else if(!shorted)
+		return ..()
+	return UI_CLOSE
+
+
+/obj/machinery/airalarm/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "AirAlarm", name)
+		ui.open()
+
+/obj/machinery/airalarm/ui_static_data(mob/user)
+	var/list/data = list()
+	data["thresholdTypeMap"] = list(
+		"warning_min" = TLV_VAR_WARNING_MIN,
+		"hazard_min" = TLV_VAR_HAZARD_MIN,
+		"warning_max" = TLV_VAR_WARNING_MAX,
+		"hazard_max" = TLV_VAR_HAZARD_MAX,
+		"all" = TLV_VAR_ALL,
+	)
+	return data
+
+/obj/machinery/airalarm/ui_data(mob/user)
+	var/data = list()
+
+	data["locked"] = locked
+	data["siliconUser"] = user.has_unlimited_silicon_privilege
+	data["emagged"] = (obj_flags & EMAGGED ? 1 : 0)
+	data["dangerLevel"] = danger_level
+	data["atmosAlarm"] = !!my_area.active_alarms[ALARM_ATMOS]
+	data["fireAlarm"] = my_area.fire
+
+	var/turf/turf = get_turf(src)
+	var/datum/gas_mixture/environment = turf.return_air()
+	var/total_moles = environment.total_moles()
+	var/temp = environment.temperature
+	var/pressure = total_moles * R_IDEAL_GAS_EQUATION * temp / environment.volume
+
+	data["envData"] = list()
+	data["envData"] += list(list(
+		"name" = "Pressure",
+		"value" = "[round(pressure, 0.01)] kPa",
+		"danger" = tlv_collection["pressure"].check_value(pressure)
+	))
+	data["envData"] += list(list(
+		"name" = "Temperature",
+		"value" = "[round(temp, 0.01)] Kelvin / [round(temp, 0.01) - T0C] Celcius",
+		"danger" = tlv_collection["temperature"].check_value(temp),
+	))
+	for(var/gas_path in environment.gases)
+		var/moles = environment.gases[gas_path][MOLES]
+		var/portion = moles / total_moles
+		data["envData"] += list(list(
+			"name" = GLOB.meta_gas_info[gas_path][META_GAS_NAME],
+			"value" = "[round(moles, 0.01)] moles / [round(100 * portion, 0.01)] % / [round(portion * pressure, 0.01)] kPa",
+			"danger" = tlv_collection[gas_path].check_value(portion * pressure),
+		))
+
+	data["tlvSettings"] = list()
+	for(var/threshold in tlv_collection)
+		var/datum/tlv/tlv = tlv_collection[threshold]
+		var/list/singular_tlv = list()
+		if(threshold == "pressure")
+			singular_tlv += list(
+				"name" = "Pressure",
+				"unit" = "kPa",
+			)
+		else if (threshold == "temperature")
+			singular_tlv += list(
+				"name" = "Temperature",
+				"unit" = "K",
+			)
+		else
+			singular_tlv += list(
+				"name" = GLOB.meta_gas_info[threshold][META_GAS_NAME],
+				"unit" = "kPa",
+			)
```

Comment:
so the assumption here is gas is the only acceptable third option? could you be explicit about that?

Reply:
also why do the list() thing? guess it's a bit easier to read then setting each var individually but I've not seen it before

Reply:
> also why do the list() thing?
oh shit right thats kinda dumb

Reply:
why not a final else if instead of a trailing else? oh right the thing's a fucking gas id god damn it.
you mind leaving a comment on it by the else?

Reply:
yeah sure

Reply:
Nah its a gaspath, we can totally do `in subtypesof()` if you want that extra safety.

Reply:
it's ok, just comment it explicitly so readers "get" it

Reply:
that or just check the truthyness of meta_gas_info
yeah do that I like that better